### PR TITLE
bpf: replace arch-specific pt_regs assumptions with bpf co-re macros

### PR DESCRIPTION
**Pull request by Forsworns**
1. add libbpf header bpf_tracing.h
2. replace probe_read to CORE macros on pt_regs
3. add a new configuration item in bpf/Makefile to assign arch
4. revise the definition `__VMLINUX_` in vmlinux.h to `__VMLINUX_H__`, which is the standard definition generated via bpftool

Signed-off-by: Peihao Yang <peihao.young@gmail.com>

### CODE REVIEW

**Comment by Forsworns** [comment]
@jrfastab I'm not sure if this is what you want.

The original `probe_read` on pt_regs are replaced by `PT_REGS_*_CORE` and `PT_REGS_*_CORE_SYSCALL`. And I have run the Tetragon locally to verify them.

But now we have to declare the `ARCH` before `make` (see changes in `bpf/Makefile`). By default, it's x86. 
Moreover, we have to regenerate the `vmlinux.h` for the specific arch, e.g., arm/arm64.

**Comment by Forsworns** [comment]
There are some "Missing or malformed SPDX-License-Identifier tag in line 1" errors in the **Check pull request commits**. But I don't know which license should be added to these headers. :)

e.g. `/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */`

**Comment by jrfastab** [comment]
Thanks @Forsworns I'll review this tomorrow.

File: bpf/lib/bpf_helpers.h
Comment (Forsworns) [reply bpf/lib/bpf_helpers.h:80]: similar to the answer in #214

File: bpf/process/bpf_fork.c
Comment (Forsworns) [reply bpf/process/bpf_fork.c:25]: I'll check it later :)

**Comment by Forsworns** [comment]
The CI checkpatch provides some suggestions on `bpf_tracing.h`, so it fails.

But they are not applicable, e.g. `CHECK:MACRO_ARG_REUSE: Macro argument reuse 'name' - possible side-effects?` for the `BPF_KPROBE` 

Maybe we can simply neglect them...

**Comment by tpapagian** [comment]
Hey @Forsworns, let me provide some thoughts on this.

First, regarding the header, instead of `/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */` please use: 
```
// SPDX-License-Identifier: GPL-2.0
/* Copyright Authors of Cilium */
```
similar to what other files have (i.e. `bpf/process/bpf_generic_kprobe.c`).

Regarding `bpf_tracing.h`, I have seen that you made changes there described in https://github.com/cilium/tetragon/pull/209/commits/6af3f7f3b50e4883f4bf45a9cd6b8ec3fccf24b4. I propose to get a copy of `bpf_tracing.h` from https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h without any changes at all (also no formatting; you can exclude that [here](https://github.com/cilium/tetragon/blob/4bd5619a17c7294824debf3a1c72901972261265/Makefile#L8) and [here](https://github.com/cilium/tetragon/blob/4bd5619a17c7294824debf3a1c72901972261265/.github/workflows/static-checks.yaml#L57)). Even the definition of `__always_inline` can be on another file. 

In the commit message you also mention: `revise kprobes' signature with co-re macros`. I believe that there is no need to do that. The `*_CORE` variants are required only for some tracepoints (check [here](https://lore.kernel.org/bpf/20200313172336.1879637-4-andriin@fb.com/) and [here](https://lore.kernel.org/bpf/20200313172336.1879637-5-andriin@fb.com/)). Here you use them only in kprobes.

The reason that I am saying that is that we don't introduce `bpf_tracing.h` only for these cases. There are several macros there that can be used in other places as well (like new features) and I think that it is not a good idea to have different semantics compared to the original file.

Finally, the use of these macros in the other files, seems fine to me.

**Comment by tpapagian** [comment]
Sorry @Forsworns , I maybe misread the comment in you commit. I created a diff of [this](https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h) and [this](https://github.com/cilium/tetragon/blob/6af3f7f3b50e4883f4bf45a9cd6b8ec3fccf24b4/bpf/libbpf/bpf_tracing.h) but still there are some differences. 

For example [this](https://github.com/libbpf/libbpf/blob/b78c75fcb347b06c31996860353f40087ed02f48/src/bpf_tracing.h#L242-L257) and [this](https://github.com/libbpf/libbpf/blob/b78c75fcb347b06c31996860353f40087ed02f48/src/bpf_tracing.h#L35-L63) are missing from the file that you have submitted. I can understand that maybe we don't care about these but I am just trying to understand why there are differences (do we need the in order to compile?). If they are needed in order to compile we are fine.

File: bpf/process/generic_calls.h
Comment (tpapagian) [inline bpf/process/generic_calls.h:116]: `BPF_KPROBE` as shown [here](https://github.com/libbpf/libbpf/blob/b78c75fcb347b06c31996860353f40087ed02f48/src/bpf_tracing.h#L441-L447) does not use the `*_CORE` variants (i.e. it uses `PT_REGS_PARM*`). Why did you made that change? 

ps. I agree with the `syscall` case where `BPF_KPROBE_SYSCALL` uses the `*_CORE` variants.

**Comment by Forsworns** [comment]
> For example [this](https://github.com/libbpf/libbpf/blob/b78c75fcb347b06c31996860353f40087ed02f48/src/bpf_tracing.h#L242-L257) and [this](https://github.com/libbpf/libbpf/blob/b78c75fcb347b06c31996860353f40087ed02f48/src/bpf_tracing.h#L35-L63) are missing from the file that you have submitted. I can understand that maybe we don't care about these but I am just trying to understand why there are differences (do we need the in order to compile?). If they are needed in order to compile we are fine.

The first one, yes. I missed this architecture. Maybe I deleted it unconsciously... I forget it.

The second one, we don't need it, since our target (indicated by Clang) will always be `bpf`. See this [line](https://github.com/cilium/tetragon/blob/6af3f7f3b50e4883f4bf45a9cd6b8ec3fccf24b4/bpf/Makefile#L26) in Makefile.

**Comment by Forsworns** [comment]
I replace the license in the `generic_calls.h` and `retprobe_map.h` to 
```
// SPDX-License-Identifier: GPL-2.0
/* Copyright Authors of Cilium */
```
as you mentioned. But I guess `bpf_tracing.h` should be kept the same as other Linux headers, since it is copied from the libbpf library, e.g., the 

https://github.com/cilium/tetragon/blob/4bd5619a17c7294824debf3a1c72901972261265/bpf/libbpf/bpf_core_read.h#L1

**Comment by Forsworns** [comment]
> (also no formatting; you can exclude that [here](https://github.com/cilium/tetragon/blob/4bd5619a17c7294824debf3a1c72901972261265/Makefile#L8) and [here](https://github.com/cilium/tetragon/blob/4bd5619a17c7294824debf3a1c72901972261265/.github/workflows/static-checks.yaml#L57)

The `bpf/libbpf/*` directory has been excluded and in fact it does not help. Because the CI failed on `checkpatch`, instead of the `static-checks`.

In the [`checkpatch.yaml`](https://github.com/cilium/tetragon/blob/6af3f7f3b50e4883f4bf45a9cd6b8ec3fccf24b4/.github/workflows/checkpatch.yaml), it uses a [docker image](https://quay.io/repository/cilium/cilium-checkpatch/manifest/sha256:aecdad89e06c4eef856313488f205d50a4fd9ff2f0fe9e20464d913add94e447). I think the script in the image should be revised....

File: bpf/libbpf/bpf_tracing.h
Comment (olsajiri) [inline bpf/libbpf/bpf_tracing.h:7]: this should be done in previous commit

File: bpf/libbpf/bpf_tracing.h
Comment (olsajiri) [inline bpf/libbpf/bpf_tracing.h:71]: why is this changed? should be done in previous commit if needed

File: bpf/libbpf/bpf_tracing.h
Comment (olsajiri) [inline bpf/libbpf/bpf_tracing.h:335]: I think this needs to be in separate commit or in the initial one and explained

File: bpf/libbpf/bpf_tracing.h
Comment (olsajiri) [inline bpf/libbpf/bpf_tracing.h:392]: I dont see this macro being used, please drop it

File: bpf/process/generic_calls.h
Comment (olsajiri) [inline bpf/process/generic_calls.h:3]: you changed this in previous comment, please change it directly to final value and make it separate commit, it will ease the review

File: bpf/process/retprobe_map.h
Comment (olsajiri) [inline bpf/process/retprobe_map.h:3]: same comment as for above header

**Review by olsajiri — COMMENTED** [review]
so what's your usecase, do you need tetragon to be built on another arch or you need CORE functionality? I think we should support both, but I'm curious how you will use the change

File: bpf/libbpf/bpf_tracing.h
Comment (Forsworns) [reply bpf/libbpf/bpf_tracing.h:349]: Thanks for your review! Yes, it's better to include `bpf_tracing.h` in a single commit, I'll revise it.

___

In fact I only include the minimum necessary definitions in the first commit.

It's tpapagian's advice to include the whole file directly from the libbpf. I also think it's good because it can ease developments in the future, so that we don't need to compare the header with libbpf if more features are need.

Then the second commit is submit.

File: bpf/Makefile
Comment (Forsworns) [reply bpf/Makefile:21]: I think the include directory may need changes since we already remove the specific assumptions on x86_64 arch.

**Comment by Forsworns** [comment]
> so what's your usecase, do you need tetragon to be built on another arch or you need CORE functionality?

@olsajiri 

In fact, it's related to the issue #190

I didn't directly close the issue with this PR since it also related to a `CI lint job` as mentioned by willfindlay in #190.

**Comment by Forsworns** [comment]
I reset this PR and re-split the committed changes.

**Comment by kkourt** [comment]
@Forsworns Thanks! Some of the comments from @olsajiri are marked as resolved but it's not clear to me how they were actually resolved. Could you please reply in each of them how they were resolved?

File: bpf/process/retprobe_map.h
Comment (Forsworns) [reply bpf/process/retprobe_map.h:3]: Resolved after re-split commits

File: bpf/process/generic_calls.h
Comment (Forsworns) [reply bpf/process/generic_calls.h:3]: Resolved after re-split commits

File: bpf/libbpf/bpf_tracing.h
Comment (Forsworns) [reply bpf/libbpf/bpf_tracing.h:71]: Resolved after re-split commits

File: bpf/libbpf/bpf_tracing.h
Comment (Forsworns) [reply bpf/libbpf/bpf_tracing.h:335]: Resolved after re-split commits

**Comment by Forsworns** [comment]
@kkourt feel free to reopen the comments you feel not resolved

**Comment by kkourt** [comment]
Thanks for adding the comments!

I'd like to have an ACK from @olsajiri before we merge this, and I think a response on how each issue was resolved will help this.

File: bpf/lib/bpf_helpers.h
Comment (olsajiri) [inline bpf/lib/bpf_helpers.h:68]: why is this needed in this commit?

File: bpf/libbpf/bpf_tracing.h
Comment (olsajiri) [inline bpf/libbpf/bpf_tracing.h:1]: I mad ediff with the latest bpf-next/master:tools/lib/bpf/bpf_tracing.h and I can see some changes are missing, could you please sync?

File: bpf/include/vmlinux.h
Comment (olsajiri) [inline bpf/include/vmlinux.h:1]: could you please move all these header license changes into separate commit?

File: bpf/process/bpf_generic_retkprobe.c
Comment (olsajiri) [inline bpf/process/bpf_generic_retkprobe.c:32]: could you please call it just 'ret', 'rc_reg' seems confusing to me that it has value of the register or something related

**Review by olsajiri — COMMENTED** [review]
looks good, added few nits

File: bpf/lib/bpf_helpers.h
Comment (Forsworns) [reply bpf/lib/bpf_helpers.h:68]: Because `bpf_core_read` calls [`bpf_probe_read`](https://github.com/cilium/tetragon/blob/bf80f7097f19efa617a53b919a3cf22cbbed3bc9/bpf/libbpf/bpf_core_read.h#L117), which is not defined in `bpf_helpers.h`. In tetragon, we once used a customized macro [`_(P)`](https://github.com/cilium/tetragon/blob/bf80f7097f19efa617a53b919a3cf22cbbed3bc9/bpf/lib/bpf_helpers.h#L51) to replace it. But it is not the "libbpf way".

File: bpf/include/vmlinux.h
Comment (Forsworns) [reply bpf/include/vmlinux.h:1]: sure :)

File: bpf/libbpf/bpf_tracing.h
Comment (Forsworns) [reply bpf/libbpf/bpf_tracing.h:1]: Have synced the changes on `PT_REGS_PARM1_SYSCALL` for arm64/s390 and new `BPF_KPROBE_SYSCALL` definitions.

Others changes are necessary, though we intend to simply maintain this header as similar to mainstream as possible. 
See the discussions [here](https://github.com/cilium/tetragon/pull/209#issuecomment-1183389988) and [here](https://github.com/cilium/tetragon/pull/209#issuecomment-1183443944)

**Comment by Forsworns** [comment]
It fails on the new enhanced tests from #240....

I'll check it ASAP

File: bpf/process/bpf_generic_retkprobe.c
Comment (Forsworns) [reply bpf/process/bpf_generic_retkprobe.c]: Revised

**Comment by olsajiri** [comment]
@Forsworns could you please rebase to the latest main, it might be the reason for the 'generated files check' fail,
other than that the change looks good to me

thanks

**Comment by Forsworns** [comment]
@olsajiri Have rebased :)

**Review by olsajiri — APPROVED** [review]

**Review by kkourt — APPROVED** [review]
Thanks!

PS: checkpatch issues seem to be a file found in libbpf,.  so it does not make sense to address them.

**Comment by tixxdz** [comment]
@Forsworns thank you!

Maybe it makes sense to add a checkpatch rule to catch future direct registers usage...

cc @kkourt

**Pull request merged**
